futures-util/src/stream/buffered.rs

```diff
@@ -13,18 +14,35 @@ use stream::{Fuse, FuturesOrdered};
 /// results in the order that they were pulled out of the original stream. This
 /// is created by the `Stream::buffered` method.
 #[must_use = "streams do nothing unless polled"]
-pub struct Buffered<S: Stream> {
+pub struct Buffered<S: Stream>
+where
+    S: Stream,
+    S::Item: Future,
+{
     stream: Fuse<S>,
     queue: FuturesOrdered<S::Item>,
     max: usize,
 }
 
+impl<S> Unpin for Buffered<S>
+where
+    S: Stream,
+    S::Item: Future,
+{}
+
+impl<S: Stream> Buffered<S>
+where
+    S: Stream,
+    S::Item: Future,
+{
+    unsafe_pinned!(stream -> Fuse<S>);
+    unsafe_unpinned!(queue -> FuturesOrdered<S::Item>);
```

Comment:
Maybe use `unsafe_pinnned` for this. It's `Unpin`, so `DerefMut` works automatically, but a pinned projection let's us avoid the `PinMut::new(self.queue())` further down => Nicer code